File: 2022-2023/sourcecode/lab/lab3/QuickYogaAppJava/app/src/main/java/pl/edu/uwr/pum/quickyogaappjava/ExerciseActivity.java
package pl.edu.uwr.pum.quickyogaappjava;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.media.MediaPlayer;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.speech.tts.TextToSpeech;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.Locale;
import java.util.Objects;

public class ExerciseActivity extends AppCompatActivity implements TextToSpeech.OnInitListener {

    private ArrayList<YogaPose> yogaPosesList;
    private int exercisePosition = 0;

    private CountDownTimer timer;
    private int progressBarValue = 0;
    private boolean isYoga = false;
    private long waitTimerDuration = 5000L;
    private long yogaTimerDuration = 7000L;
    private long timerInterval = 1000;
    private int waitProgressInterval = 5;
    private int yogaProgressInterval = 7;

    private ProgressBar progressBar;
    private TextView textViewCounter;
    private ImageView imageView;
    private TextView textViewExerciseName;

    private TextToSpeech textToSpeech;
    private MediaPlayer player;

    private YogaAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_exercise);

        progressBar = findViewById(R.id.progress_bar);
        textViewCounter = findViewById(R.id.text_view_counter);
        imageView = findViewById(R.id.yoga_image);
        textViewExerciseName = findViewById(R.id.text_view_exercise_title);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);
        toolbar.setNavigationOnClickListener(view -> onBackPressed());
        textToSpeech = new TextToSpeech(this, this);
        yogaPosesList = YogaPoses.getYogaPoses();
        setupProgressBar();
        RecyclerView recyclerView = findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(
                new LinearLayoutManager(
                        this,
                        LinearLayoutManager.HORIZONTAL,
                        false));
        adapter = new YogaAdapter(this, yogaPosesList);
        recyclerView.setAdapter(adapter);

    }

    @Override
    protected void onDestroy() {
        if (timer != null) {
            timer.cancel();
            progressBarValue = 0;
        }

        if (player != null)
            player.stop();

        exercisePosition = 0;

        if (textToSpeech != null) {
            textToSpeech.stop();
            textToSpeech.shutdown();
        }

        super.onDestroy();
    }

    private void setupProgressBar(){
        long timerDuration;
        int progressInterval;
        if (isYoga){
            timerDuration = yogaTimerDuration;
            progressInterval = yogaProgressInterval;
            speakPoseName(yogaPosesList.get(exercisePosition).getName());
            progressBar.setMax(yogaProgressInterval);
            textViewExerciseName.setText(yogaPosesList.get(exercisePosition).getName());
            imageView.setVisibility(View.VISIBLE);
            yogaPosesList.get(exercisePosition).setSelected(true);
            adapter.notifyItemChanged(exercisePosition);
            imageView.setImageResource(yogaPosesList.get(exercisePosition).getImage());
        }else{
            timerDuration = waitTimerDuration;
            progressInterval = waitProgressInterval;
            if (exercisePosition > 0) {
                try {
                    player = MediaPlayer.create(this, R.raw.dingding);
                    player.setLooping(false);
                    player.start();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
            progressBar.setMax(progressInterval);
            imageView.setVisibility(View.INVISIBLE);
            textViewExerciseName.setText(getString(R.string.wait_text));
        }
        progressBar.setProgress(progressBarValue);
        timer = new CountDownTimer(timerDuration, timerInterval) {
            @Override
            public void onTick(long l) {
                progressBarValue++;
                progressBar.setProgress(progressInterval - progressBarValue);
                textViewCounter.setText(String.valueOf(progressInterval - progressBarValue));
            }

            @Override
            public void onFinish() {
                Toast.makeText(ExerciseActivity.this, "YOGA TIME", Toast.LENGTH_SHORT).show();
                if (timer != null) {
                    timer.cancel();
                    progressBarValue = 0;
                }

                if (isYoga){
                    yogaPosesList.get(exercisePosition).setSelected(false);
                    yogaPosesList.get(exercisePosition).setCompleted(true);
                    adapter.notifyItemChanged(exercisePosition);
                    exercisePosition++;
                }

                isYoga = !isYoga;

                if (exercisePosition < yogaPosesList.size())
                    setupProgressBar();
                else
                    Toast.makeText(ExerciseActivity.this, "COMPLETE", Toast.LENGTH_SHORT).show();
            }
        }.start();
    }

    @Override
    public void onInit(int i) {
        if (i == TextToSpeech.SUCCESS)
            textToSpeech.setLanguage(Locale.ENGLISH);
    }

    private void speakPoseName(String text){
        textToSpeech.speak(text, TextToSpeech.QUEUE_FLUSH, null, "");
    }
}

File: 2022-2023/sourcecode/lab/lab3/QuickYogaAppKotlin/app/src/main/java/pl/edu/uwr/pum/quickyogaappkotlin/ExerciseActivity.kt
package pl.edu.uwr.pum.quickyogaappkotlin

import android.media.MediaPlayer
import android.os.Bundle
import android.os.CountDownTimer
import android.speech.tts.TextToSpeech
import android.view.View
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import java.util.*
import kotlin.collections.ArrayList


class ExerciseActivity : AppCompatActivity(), TextToSpeech.OnInitListener {

    private lateinit var timer: CountDownTimer

    private var progressBarValue = 0
    private val waitTimerDuration = 5000L
    private val timerInterval: Long = 1000
    private val waitProgressInterval = 5
    private val yogaTimerDuration = 7000L
    private val yogaProgressInterval = 7

    private var isYoga = false

    private val yogaPosesList: ArrayList<YogaPose> = YogaPoses.yogaPoses
    private var exercisePosition = 0

    private val textToSpeech by lazy { TextToSpeech(this, this) }
    private lateinit var mediaPlayer: MediaPlayer

    private val recyclerView by lazy { findViewById<RecyclerView>(R.id.recyclerView) }
    private val adapter: YogaAdapter by lazy { YogaAdapter(this, yogaPosesList) }

    private val progressBar by lazy { findViewById<ProgressBar>(R.id.progress_bar)}
    private val textViewCounter by lazy { findViewById<TextView>(R.id.text_view_counter) }
    private val imageView by lazy { findViewById<ImageView>(R.id.yoga_image) }
    private val textViewExerciseName by lazy { findViewById<TextView>(R.id.text_view_exercise_title) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_exercise)
        val toolbar: Toolbar = findViewById(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        toolbar.setNavigationOnClickListener { onBackPressed() }
        setupProgressBar()

        recyclerView.layoutManager = LinearLayoutManager(
            this,
            LinearLayoutManager.HORIZONTAL,
            false
        )
        recyclerView.adapter = adapter
    }

    override fun onDestroy() {
        timer.cancel()
        progressBarValue = 0

        exercisePosition = 0; // resetowanie pozycji

        textToSpeech.stop();
        textToSpeech.shutdown();

        mediaPlayer.stop()

        super.onDestroy()
    }

    private fun setupProgressBar() {
        val timerDuration: Long
        val progressInterval: Int
        if (isYoga) {
            // ćwiczenie
            timerDuration = yogaTimerDuration
            progressInterval = yogaProgressInterval
            speakPoseName(yogaPosesList[exercisePosition].name)
            progressBar.max = yogaProgressInterval
            textViewExerciseName.text = yogaPosesList[exercisePosition].name // tytuł
            imageView.visibility = View.VISIBLE
            imageView.setImageResource(yogaPosesList[exercisePosition].image) // grafika
            yogaPosesList[exercisePosition].isSelected = true
            adapter.notifyItemChanged(exercisePosition)
        } else {
            // oczekiwanie
            timerDuration = waitTimerDuration
            progressInterval = waitProgressInterval
            try {
                mediaPlayer = MediaPlayer.create(this, R.raw.dingding)
                mediaPlayer.isLooping = false
                mediaPlayer.start()
            } catch (e: Exception){
                e.printStackTrace()
            }
            progressBar.max = progressInterval
            imageView.visibility = View.INVISIBLE
            textViewExerciseName.text = getString(R.string.wait_text)
        }
        timer = object: CountDownTimer(timerDuration, timerInterval) {
            override fun onTick(l: Long) {
                progressBarValue++
                progressBar.progress = progressInterval - progressBarValue
                textViewCounter.text = (progressInterval - progressBarValue).toString()}
            override fun onFinish() {
                if (exercisePosition < yogaPosesList.size)
                    setupProgressBar()
                else
                    Toast.makeText(this@ExerciseActivity, "COMPLETE", Toast.LENGTH_SHORT).show();
                timer.cancel()
                progressBarValue = 0

                if (isYoga) {
                    yogaPosesList[exercisePosition].isSelected = false
                    yogaPosesList[exercisePosition].isCompleted = true
                    adapter.notifyItemChanged(exercisePosition);
                    exercisePosition++
                }

                isYoga = !isYoga
                setupProgressBar()
            }
        }.start()
        progressBar.progress = progressBarValue
    }

    override fun onInit(p0: Int) {
        if (p0 == TextToSpeech.SUCCESS)
            textToSpeech.language = Locale.ENGLISH
    }

    private fun speakPoseName(text: String) {
        textToSpeech.speak(text, TextToSpeech.QUEUE_FLUSH, null, "")
    }
}